internal/infoschema.go

```diff
@@ -48,6 +53,150 @@ func ProcessInfoSchema(conv *Conv, db *sql.DB) error {
 	return nil
 }
 
+// ProcessSqlData performs data conversion for source database
+// 'db'. For each table, we extract data using a "SELECT *" query,
+// convert the data to Spanner data (based on the source and Spanner
+// schemas), and write it to Spanner.  If we can't get/process data
+// for a table, we skip that table and process the remaining tables.
+//
+// Note that the database/sql library has a somewhat complex model for
+// returning data from rows.Scan. Scalar values can be returned using
+// the native value used by the underlying driver (by passing
+// *interface{} to rows.Scan), or they can be converted to specific go
+// types. Array values are always returned as []byte, a string
+// encoding of the array values. This string encoding is
+// database/driver specific. For example, for PostgreSQL, array values
+// are returned in the form "{v1,v2,..,vn}", where each v1,v2,...,vn
+// is a PostgreSQL encoding of the respective array value.
+//
+// We choose to do all type conversions explicitly ourselves so that
+// we can generate more targeted error messages: hence we pass
+// *interface{} parameters to row.Scan.
+func ProcessSqlData(conv *Conv, db *sql.DB) {
+	// TODO: refactor to use the set of tables computed by
+	// ProcessInfoSchema instead of computing them again.
+	tables, err := getTables(db)
+	if err != nil {
+		conv.unexpected(fmt.Sprintf("Couldn't get list of table: %s", err))
+		return
+	}
+	for _, t := range tables {
+		// PostgreSQL schema and name can be arbitrary strings.
+		// Ideally we would pass schema/name as a query parameter,
+		// but PostgreSQL doesn't support this. So we quote it instead.
+		q := fmt.Sprintf(`SELECT * FROM "%s"."%s";`, t.schema, t.name)
+		rows, err := db.Query(q)
+		if err != nil {
+			conv.unexpected(fmt.Sprintf("Couldn't get data for table: %s", err))
+			continue
+		}
+		defer rows.Close()
+		srcTable := buildTableName(t.schema, t.name)
+		srcCols, err1 := rows.Columns()
+		spTable, err2 := GetSpannerTable(conv, srcTable)
+		spCols, err3 := GetSpannerCols(conv, srcTable, srcCols)
+		spSchema, ok1 := conv.spSchema[spTable]
+		srcSchema, ok2 := conv.srcSchema[srcTable]
+		if err1 != nil || err2 != nil || err3 != nil || !ok1 || !ok2 {
+			conv.statsAddBadRows(srcTable, conv.stats.rows[srcTable])
+			conv.unexpected(fmt.Sprintf("Can't get cols and schemas for table %s: err1=%s, err2=%s, err3=%s, ok1=%t, ok2=%t",
+				srcTable, err1, err2, err3, ok1, ok2))
+			continue
+		}
+		v, iv := buildVals(len(srcCols))
+		for rows.Next() {
+			err := rows.Scan(iv...)
+			if err != nil {
+				conv.unexpected(fmt.Sprintf("Couldn't process sql data row: %s", err))
+				// Scan failed, so we don't have any data to add to bad rows.
+				conv.statsAddBadRow(srcTable, conv.dataMode())
+				continue
+			}
+			cvtCols, cvtVals, err := ConvertSqlRow(conv, srcTable, srcCols, srcSchema, spTable, spCols, spSchema, v)
```

Comment:
This is discussed in the documentation for ConvertSqlRow.
In short, cvtCols is a subset of spCols: cols can change when we add a column (synthetic primary key) or because we drop columns (handling of NULL values).

Reply:
The other option would be to pull the synthetic-primary key stuff out of CollectBadRows, and handle NULL values using nil.
But ProcessSqlData is already a large function.
Let's leave this till we address the bigger refactoring issue i.e. using a transaction for all of the schema and data queries.